src/tss2-tcti/tcti-helper-common.c

```diff
@@ -0,0 +1,761 @@
+        return rc;
+    }
+
+    LOG_DEBUG ("TPM is ready");
+
+    /* Get rid */
+    rc = ctx->read_reg (ctx->data, TCTI_HELPER_COMMON_REG_TPM_RID, &rid, sizeof(rid));
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to read the TPM_RID register");
+        return rc;
+    }
+
+#if MAXLOGLEVEL >= LOGL_INFO
+    /* Print device details */
+    uint16_t vendor_id, device_id;
+    vendor_id = did_vid & 0xffff;
+    device_id = did_vid >> 16;
+    LOG_DEBUG ("Connected to TPM with vid:did:rid of 0x%04" PRIx16 ":%04" PRIx16 ":%02" PRIx8, vendor_id, device_id, rid);
+#endif
+
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC Tcti_Helper_Common_Transmit (TCTI_HELPER_COMMON_CONTEXT *ctx, size_t size, const uint8_t *cmd_buf)
+{
+    TSS2_RC rc;
+    tpm_header_t header;
+    uint32_t expected_status_bits;
+
+    if (ctx == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+
+    rc = header_unmarshal (cmd_buf, &header);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    if (header.size != size) {
+        LOG_ERROR ("Buffer size parameter: %zu, and TPM2 command header size "
+                   "field: %" PRIu32 " disagree.", size, header.size);
+        return TSS2_TCTI_RC_BAD_VALUE;
+    }
+
+    LOGBLOB_DEBUG (cmd_buf, size, "Sending command with TPM_CC 0x%08"PRIx32" and size %" PRIu32,
+               header.code, header.size);
+
+    /* Tell TPM to expect command */
+    rc = tcti_helper_common_write_sts_reg (ctx, TCTI_HELPER_COMMON_TPM_STS_COMMAND_READY);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to write the commandReady bit to TPM_STS register");
+        return rc;
+    }
+
+    /* Wait until ready bit is set by TPM device */
+    expected_status_bits = TCTI_HELPER_COMMON_TPM_STS_COMMAND_READY;
+    rc = tcti_helper_common_wait_for_status (ctx, expected_status_bits, expected_status_bits, TIMEOUT_B);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to wait for the TPM to become ready (commandReady == 1)");
+        return rc;
+    }
+
+    /* Send the first byte of the command */
+    rc = tcti_helper_common_fifo_transfer (ctx, (void *)cmd_buf, 1, TCTI_HELPER_COMMON_FIFO_TRANSMIT_1);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to send the command");
+        return rc;
+    }
+
+    /* Send all but the last byte in the FIFO */
+    rc = tcti_helper_common_fifo_transfer (ctx, (void *)(cmd_buf + 1), size - 2, TCTI_HELPER_COMMON_FIFO_TRANSMIT_2);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to send the command");
+        return rc;
+    }
+
+    /* Send the last byte */
+    rc = tcti_helper_common_fifo_transfer (ctx, (void *)(cmd_buf + size - 1), 1, TCTI_HELPER_COMMON_FIFO_TRANSMIT_2);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to send the command");
+        return rc;
+    }
+
+    /* Await TPM readiness */
+    expected_status_bits = TCTI_HELPER_COMMON_TPM_STS_VALID | TCTI_HELPER_COMMON_TPM_STS_DATA_EXPECT;
```

Comment:
Here, the expected status should be STS_VALID | !EXPECT

I wonder why this still works ?